gc/base/standard/ConcurrentGCSATB.cpp

```diff
@@ -142,4 +176,240 @@ MM_ConcurrentGCSATB::localMark(MM_EnvironmentBase *env, uintptr_t sizeToTrace)
 	return sizeTraced;
 }
 
+#if defined(OMR_GC_MODRON_SCAVENGER)
+/**
+ * Process event from an external GC (Scavenger) when old-to-old reference is created.
+ * @param objectPtr  Parent old object that has a reference to a child old object
+ */
+void
+MM_ConcurrentGCSATB::oldToOldReferenceCreated(MM_EnvironmentBase *env, omrobjectptr_t objectPtr)
+{}
+#endif /* OMR_GC_MODRON_SCAVENGER */
+
+
+/**
+ * Tune the concurrent adaptive parameters.
+ * Using historical data attempt to predict how much work (tracing)
+ * will need to be performed during the next concurrent mark cycle.
+ */
+void
+MM_ConcurrentGCSATB::tuneToHeap(MM_EnvironmentBase *env)

```

Comment:
without me drilling into the code diffs, could you explain how is this conceptually different from IncrementaUpdate?

Reply:
**SATB borrows logic from Incremental and simplifies it. Conceptually it is the pretty much same as Incremental:**

1. For SATB, `_bytesToTracePass1`/`_bytesToTracePass2`/`_bytesToCleanPass1`/`_bytesToCleanPass2` are simplified  to `_bytesToTrace`, for example:

```C++
totalBytesToTrace = (uintptr_t)(heapSize * _tenureLiveObjectFactor * _tenureNonLeafObjectFactor);
_bytesToTracePass1 = (uintptr_t)((float)totalBytesToTrace * _bytesTracedInPass1Factor);
_bytesToTracePass2 = MM_Math::saturatingSubtract(totalBytesToTrace, _bytesToTracePass1);
 _bytesToCleanPass1 = (uintptr_t)((float)totalBytesToTrace * _cardCleaningFactorPass1);
 _bytesToCleanPass2 = (uintptr_t)((float)totalBytesToTrace * _cardCleaningFactorPass2);
```
is simplified to:
```C++
_bytesToTrace = (uintptr_t)(heapSize * _tenureLiveObjectFactor * _tenureNonLeafObjectFactor);
```


2. Both configs tuning is branched based on the same conditions:

```C++
if(0 == _stats.getKickoffThreshold() || _retuneAfterHeapResize ) {
	... /* We estimate new bytes to trace based on average live rate. */
} else {
	/* Re-tune based on actual amount traced if we completed tracing on last cycle */
	if ((NULL == env->_cycleState) || env->_cycleState->_gcCode.isExplicitGC() || !_stwCollectionInProgress) {
		/* Nothing to do - we can't update statistics on a system GC or when no cycle is running */
	} else if (CONCURRENT_EXHAUSTED <= _stats.getExecutionModeAtGC()) {
              ... /* weighted average of total traced and previous bytes to trace */
	} else if (_stats.getExecutionModeAtGC() == CONCURRENT_TRACE_ONLY) {
		/* Assume amount to be traced on next cycle will what we traced this time PLUS
		 * the tracing we did to complete processing of any work packets that remained at
		 * the start of the collection.
		 * This is an over estimate but will get us back on track.
		 */
		...
	} else {
		/* We did not trace enough to use amount traced to predict trace target so best we can do
		 * is estimate based on current heap size, live object factor, leaf object factor etc.
		 */
		...
	}
}

```
The difference here is that when we don't complete Concurrent (when state <= EXHAUSTED), `Incremental` will only consider sufficient progress for estimation _(2nd if else case)_ if CM makes it to `CLEAN_TRACE`. Whereas SATB will estimate at `TRACE_ONLY`

3. The other differences are subtle, they are a result of simplification of the params and accounting for only 1.. e.g.,:
```C++
kickoffThreshold = (_stats.getInitWorkRequired() / _allocToInitRate) +
					   (_bytesToTrace / _allocToTraceRateNormal);
```
vs
```C++
kickoffThreshold = (_stats.getInitWorkRequired() / _allocToInitRate) +
					   (_traceTargetPass1 / _allocToTraceRateNormal) +
					   (_traceTargetPass2 / (_allocToTraceRateNormal * _allocToTraceRateCardCleanPass2Boost));
```